plugins/woocommerce/src/Internal/ProductQueryFilters/FilterClausesGenerator.php

```diff
@@ -1,254 +1,202 @@
-
-		$attributes_to_count_sql = 'AND term_taxonomy.taxonomy IN ("' . esc_sql( wc_sanitize_taxonomy_name( $attribute_to_count ) ) . '")';
-		$attribute_count_sql     = "
-			SELECT COUNT( DISTINCT posts.ID ) as term_count, terms.term_id as term_count_id
-			FROM {$wpdb->posts} AS posts
-			INNER JOIN {$wpdb->term_relationships} AS term_relationships ON posts.ID = term_relationships.object_id
-			INNER JOIN {$wpdb->term_taxonomy} AS term_taxonomy USING( term_taxonomy_id )
-			INNER JOIN {$wpdb->terms} AS terms USING( term_id )
-			WHERE posts.ID IN ( {$product_query_sql} )
-			{$attributes_to_count_sql}
-			GROUP BY terms.term_id
-		";
-
-		$results = $wpdb->get_results( $attribute_count_sql ); // phpcs:ignore
-
-		return array_map( 'absint', wp_list_pluck( $results, 'term_count', 'term_count_id' ) );
-	}
-
-	/**
-	 * Add query clauses for stock filter.
+	 * @param array $args              Query args.
+	 * @param array $chosen_attributes {
+	 *     Chosen attributes array.
 	 *
-	 * @param array     $args     Query args.
-	 * @param \WP_Query $wp_query WP_Query object.
+	 *     @type array {$taxonomy: Attribute taxonomy name} {
+	 *         @type string[] $terms      Chosen terms' slug.
+	 *         @type string   $query_type Query type. Accepts 'and' or 'or'.
+	 *     }
+	 * }
 	 * @return array
 	 */
-	private function stock_filter_clauses( $args, $wp_query ) {
-		if ( ! $wp_query->get( 'filter_stock_status' ) ) {
+	public function add_attribute_clauses( $args, $chosen_attributes ) {
+		if ( empty( $chosen_attributes ) ) {
 			return $args;
 		}
 
-		$args['join']   = $this->append_product_sorting_table_join( $args['join'] );
-		$args['where'] .= ' AND wc_product_meta_lookup.stock_status IN ("' . implode( '","', array_map( 'esc_sql', explode( ',', $wp_query->get( 'filter_stock_status' ) ) ) ) . '")';
-
-		return $args;
-	}
+		global $wpdb;
 
-	/**
-	 * Add query clauses for price filter.
-	 *
-	 * @param array     $args     Query args.
-	 * @param \WP_Query $wp_query WP_Query object.
-	 * @return array
-	 */
-	private function price_filter_clauses( $args, $wp_query ) {
-		if ( ! $wp_query->get( 'min_price' ) && ! $wp_query->get( 'max_price' ) ) {
-			return $args;
+		// The extra derived table ("SELECT product_or_parent_id FROM") is needed for performance
+		// (causes the filtering subquery to be executed only once).
+		$clause_root = " {$wpdb->posts}.ID IN ( SELECT product_or_parent_id FROM (";
+		if ( 'yes' === get_option( 'woocommerce_hide_out_of_stock_items' ) ) {
+			$in_stock_clause = ' AND in_stock = 1';
+		} else {
+			$in_stock_clause = '';
 		}
 
-		global $wpdb;
+		$attribute_ids_for_and_filtering = array();
 
-		$adjust_for_taxes = $this->adjust_price_filters_for_displayed_taxes();
-		$args['join']     = $this->append_product_sorting_table_join( $args['join'] );
+		foreach ( $chosen_attributes as $taxonomy => $data ) {
+			$all_terms                  = get_terms( $taxonomy, array( 'hide_empty' => false ) );
```

Comment:
`get_terms` inside a `foreach` is not great. Can we cache this? or figure out a bulk query, preferably both? Additionally, should this be moved to `FilterDataProvider`?

Reply:
> get_terms inside a foreach is not great. Can we cache this? or figure out a bulk query, preferably both?

Good point! Thanks. I believe we can do both. We can also apply that fix to https://github.com/woocommerce/woocommerce/blob/3f0cf66ba050a2258e79a03590bf7eaba9343ae5/plugins/woocommerce/src/Internal/ProductAttributesLookup/Filterer.php#L232-L240.

> Additionally, should this be moved to FilterDataProvider?

Can you explain why you think so? We need selected attribute terms to modify the query clause only, the attribute count doesn't need selected terms data.

Reply:
I addressed this in https://github.com/woocommerce/woocommerce/pull/43772/commits/98a957513f3df1193a6103dc04643c195e86c930